File: 494-target-sum/494-target-sum.java
class Solution {
    int ans = 0;
    public int findTargetSumWays(int[] nums, int target) {
        helper(nums,target,0,0);
        return ans;
    }
    
    private void helper(int[] nums, int target,int sum, int idx){
        if(idx == nums.length){
            ans += (target == sum) ? 1 : 0;
            return;
        }
        
        helper(nums,target, sum + nums[idx], idx + 1);
        helper(nums,target,sum - nums[idx], idx + 1);
    }
}

/*
    [1,1,1,1,1], target = 3;
    
    


*/

File: 494-target-sum/494-target-sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        total = sum(nums)
        if(abs(target) > total):
            return 0
        n = len(nums)
        dp = [[0 for _ in range(2*total + 1)] for _ in range(n+1)]
        dp[0][total] = 1
        
        for i in range(n):
            for t in range(-1*total,total + 1, 1):
                prev = dp[i][total + t-nums[i]] if t - nums[i] >= -1*total else 0
                nxt = dp[i][total + t + nums[i]] if t + nums[i] <= total else 0
                
                dp[i+1][total + t] = prev  + nxt
                
        return dp[n][total + target]